File: tests/UnitTestApp/UnitTests/ANGDPRSettingsTestCase.m
/*   Copyright 2020 APPNEXUS INC

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

#import <XCTest/XCTest.h>
#import "ANGDPRSettings.h"

NSString * const  ANGDPR_ConsentString = @"ANGDPR_ConsentString";
NSString * const  ANGDPR_ConsentRequired = @"ANGDPR_ConsentRequired";
NSString * const  ANGDPR_PurposeConsents = @"ANGDPR_PurposeConsents";

//TCF 2.0 variables
NSString * const  ANIABTCF_ConsentString = @"IABTCF_TCString";
NSString * const  ANIABTCF_SubjectToGDPR = @"IABTCF_gdprApplies";
NSString * const  ANIABTCF_PurposeConsents = @"IABTCF_PurposeConsents";
// Gpp TCF 2.0 variabled
NSString * const  ANIABGPP_TCFEU2_PurposeConsents = @"IABGPP_TCFEU2_PurposeConsents";
NSString * const  ANIABGPP_TCFEU2_SubjectToGDPR = @"IABGPP_TCFEU2_gdprApplies";

//TCF 1.1 variables
NSString * const  ANIABConsent_ConsentString = @"IABConsent_ConsentString";
NSString * const  ANIABConsent_SubjectToGDPR = @"IABConsent_SubjectToGDPR";

@interface ANGDPRSettingsTestCase : XCTestCase

@end

@implementation ANGDPRSettingsTestCase

- (void)setUp {
    // Put setup code here. This method is called before the invocation of each test method in the class.
}

- (void)tearDown {
    [ANGDPRSettings reset];
    [[NSUserDefaults standardUserDefaults] removeObjectForKey:ANIABTCF_ConsentString];
    [[NSUserDefaults standardUserDefaults] removeObjectForKey:ANIABConsent_ConsentString];
    [[NSUserDefaults standardUserDefaults] removeObjectForKey:ANIABTCF_SubjectToGDPR];
    [[NSUserDefaults standardUserDefaults] removeObjectForKey:ANIABConsent_SubjectToGDPR];
    [[NSUserDefaults standardUserDefaults] removeObjectForKey:ANIABTCF_PurposeConsents];
    [[NSUserDefaults standardUserDefaults] removeObjectForKey:ANIABGPP_TCFEU2_PurposeConsents];
    [[NSUserDefaults standardUserDefaults] removeObjectForKey:ANIABGPP_TCFEU2_SubjectToGDPR];
    
}

- (void)testGDPRConsentString
{
    [ANGDPRSettings setConsentString:@"BOMyQRvOMyQRvABABBAAABAAAAAAEA"];
    XCTAssertTrue([[ANGDPRSettings getConsentString] isEqualToString:@"BOMyQRvOMyQRvABABBAAABAAAAAAEA"]);
}

- (void)testGDPRConsentStringWithEmpty
{
    [ANGDPRSettings setConsentString:@""];
    XCTAssertTrue([[ANGDPRSettings getConsentString] isEqualToString:@""]);
}

- (void)testIABTCFConsentString
{
    [[NSUserDefaults standardUserDefaults] setObject:@"BOMyQRvOMyQRvABABBAAABAAAAAAEA" forKey:ANIABTCF_ConsentString];
    XCTAssertTrue([[ANGDPRSettings getConsentString] isEqualToString:@"BOMyQRvOMyQRvABABBAAABAAAAAAEA"]);
}

- (void)testIABTCFConsentStringWithEmpty
{
    [[NSUserDefaults standardUserDefaults] setObject:@"" forKey:ANIABTCF_ConsentString];
    XCTAssertTrue([[ANGDPRSettings getConsentString] isEqualToString:@""]);
}

- (void)testIABConsentString
{
    [[NSUserDefaults standardUserDefaults] setObject:@"BOMyQRvOMyQRvABABBAAABAAAAAAEA" forKey:ANIABConsent_ConsentString];
    XCTAssertTrue([[ANGDPRSettings getConsentString] isEqualToString:@"BOMyQRvOMyQRvABABBAAABAAAAAAEA"]);
}

- (void)testIABConsentStringWithEmpty
{
    [[NSUserDefaults standardUserDefaults] setObject:@"" forKey:ANIABConsent_ConsentString];
    XCTAssertTrue([[ANGDPRSettings getConsentString] isEqualToString:@""]);
}

- (void)testGDPRConsentRequiredTrue
{
    [ANGDPRSettings setConsentRequired:[NSNumber numberWithInt:1]];
    XCTAssertTrue([[ANGDPRSettings getConsentRequired] boolValue] == true);
}

- (void)testGDPRConsentRequiredFalse
{
    [ANGDPRSettings setConsentRequired:[NSNumber numberWithInt:0]];
    XCTAssertTrue([[ANGDPRSettings getConsentRequired] boolValue] == false);
}

- (void)testIABTCFConsentRequiredTrue
{
    [[NSUserDefaults standardUserDefaults] setValue:[NSNumber numberWithInt:1] forKey:ANIABTCF_SubjectToGDPR];
    XCTAssertTrue([[ANGDPRSettings getConsentRequired] boolValue] == true);
}

- (void)testIABTCFConsentRequiredFalse
{
    [[NSUserDefaults standardUserDefaults] setValue:[NSNumber numberWithInt:0] forKey:ANIABTCF_SubjectToGDPR];
    XCTAssertTrue([[ANGDPRSettings getConsentRequired] boolValue] == false);
}

- (void)testIABConsentRequiredTrue
{
    [[NSUserDefaults standardUserDefaults] setValue:[NSNumber numberWithInt:1] forKey:ANIABConsent_SubjectToGDPR];
    XCTAssertTrue([[ANGDPRSettings getConsentRequired] boolValue] == true);
}

- (void)testIABConsentRequiredFalse
{
    [[NSUserDefaults standardUserDefaults] setValue:[NSNumber numberWithInt:0] forKey:ANIABConsent_SubjectToGDPR];
    XCTAssertTrue([[ANGDPRSettings getConsentRequired] boolValue] == false);
}

- (void)testGDPRPurposeConsents
{
    [ANGDPRSettings setPurposeConsents:@"10101001"];
    XCTAssertTrue([[ANGDPRSettings getDeviceAccessConsent] isEqualToString:@"1"]);
}

- (void)testGDPRPurposeConsentsEmpty
{
    [ANGDPRSettings setPurposeConsents:@""];
    XCTAssertTrue([ANGDPRSettings getDeviceAccessConsent] == nil);
}

- (void)testIABTCFPurposeConsents
{
    [[NSUserDefaults standardUserDefaults] setObject:@"10101001" forKey:ANIABTCF_PurposeConsents];
    XCTAssertTrue([[ANGDPRSettings getDeviceAccessConsent] isEqualToString:@"1"]);
}

- (void)testIABTCFPurposeConsentsEmpty
{
    [[NSUserDefaults standardUserDefaults] setObject:@"" forKey:ANIABTCF_PurposeConsents];
    XCTAssertTrue([ANGDPRSettings getDeviceAccessConsent] == nil);
}

- (void)testAccessDeviceData_PurposeConsentsEmpty_ConsentRequiredEmpty
{
    XCTAssertTrue([ANGDPRSettings canAccessDeviceData] == true);
}

- (void)testAccessDeviceData_PurposeConsentsEmpty_ConsentRequiredFalse
{
    [ANGDPRSettings setConsentRequired:[NSNumber numberWithInt:0]];
    XCTAssertTrue([ANGDPRSettings canAccessDeviceData] == true);
}

- (void)testAccessDeviceData_PurposeConsentsEmpty_IABTCFConsentRequiredFalse
{
    [[NSUserDefaults standardUserDefaults] setValue:[NSNumber numberWithInt:0] forKey:ANIABTCF_SubjectToGDPR];
    XCTAssertTrue([ANGDPRSettings canAccessDeviceData] == true);
}

- (void)testAccessDeviceData_PurposeConsentsEmpty_IABConsentRequiredFalse
{
    [[NSUserDefaults standardUserDefaults] setValue:[NSNumber numberWithInt:0] forKey:ANIABConsent_SubjectToGDPR];
    XCTAssertTrue([ANGDPRSettings canAccessDeviceData] == true);
}

- (void)testAccessDeviceData_PurposeConsentsEmpty_ConsentRequiredTrue
{
    [ANGDPRSettings setConsentRequired:[NSNumber numberWithInt:1]];
    XCTAssertTrue([ANGDPRSettings canAccessDeviceData] == false);
}

- (void)testAccessDeviceData_PurposeConsentsEmpty_IABTCFConsentRequiredTrue
{
    [[NSUserDefaults standardUserDefaults] setValue:[NSNumber numberWithInt:1] forKey:ANIABTCF_SubjectToGDPR];
    XCTAssertTrue([ANGDPRSettings canAccessDeviceData] == false);
}

- (void)testAccessDeviceData_PurposeConsentsEmpty_IABConsentRequiredTrue
{
    [[NSUserDefaults standardUserDefaults] setValue:[NSNumber numberWithInt:1] forKey:ANIABConsent_SubjectToGDPR];
    XCTAssertTrue([ANGDPRSettings canAccessDeviceData] == false);
}

- (void)testAccessDeviceData_GDPR_PurposeConsents_True
{
   [ANGDPRSettings setPurposeConsents:@"10101001"];
    XCTAssertTrue([ANGDPRSettings canAccessDeviceData] == true);
}

- (void)testAccessDeviceData_IABTCF_PurposeConsents_True
{
    [[NSUserDefaults standardUserDefaults] setObject:@"10101001" forKey:ANIABTCF_PurposeConsents];
    XCTAssertTrue([ANGDPRSettings canAccessDeviceData] == true);
}

- (void)testAccessDeviceData_GDPR_PurposeConsents_False
{
   [ANGDPRSettings setPurposeConsents:@"01010110"];
    XCTAssertTrue([ANGDPRSettings canAccessDeviceData] == false);
}

- (void)testAccessDeviceData_IABTCF_PurposeConsents_False
{
    [[NSUserDefaults standardUserDefaults] setObject:@"01010110" forKey:ANIABTCF_PurposeConsents];
    XCTAssertTrue([ANGDPRSettings canAccessDeviceData] == false);
}


- (void)testAccessDeviceData_IABGPP_TCFEU2_PurposeConsents_True
{
    [[NSUserDefaults standardUserDefaults] setObject:@"10101001" forKey:ANIABGPP_TCFEU2_PurposeConsents];
    XCTAssertTrue([ANGDPRSettings canAccessDeviceData] == true);
}


- (void)testAccessDeviceData_IABGPP_TCFEU2_PurposeConsents_False
{
    [[NSUserDefaults standardUserDefaults] setObject:@"01010110" forKey:ANIABGPP_TCFEU2_PurposeConsents];
    XCTAssertTrue([ANGDPRSettings canAccessDeviceData] == false);
}

@end

File: SwiftCompatibilityTest/SwiftCompatibilityTestTests/ANGDPRSettings/ANGDPRSettingsTestCase.swift
/*
 *
 *    Copyright 2018 APPNEXUS INC
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

import XCTest
import AppNexusSDK

class ANGDPRSettingsTestCase: XCTestCase, ANBannerAdViewDelegate {
    
    var banner : ANBannerAdView!
    weak var loadAdSuccesfulException : XCTestExpectation?
    var timeoutForImpbusRequest: TimeInterval = 0.0
    private var placementID = "4019246"
    var request: URLRequest!
    var jsonRequestBody = [String : Any]()
    
    override func setUp() {
        super.setUp()
        // Put setup code here. This method is called before the invocation of each test method in the class.
        banner = nil
        timeoutForImpbusRequest = 10.0
        ANHTTPStubbingManager.shared().enable()
        ANHTTPStubbingManager.shared().ignoreUnstubbedRequests = true
        ANHTTPStubbingManager.shared().broadcastRequests = true
        NotificationCenter.default.addObserver(self, selector: #selector(self.requestCompleted(_:)), name: NSNotification.Name.anhttpStubURLProtocolRequestDidLoad, object: nil)
        request = nil
    }
    
    override func tearDown() {
        // Put teardown code here. This method is called after the invocation of each test method in the class.
        super.tearDown()
        ANHTTPStubbingManager.shared().disable()
        ANHTTPStubbingManager.shared().removeAllStubs()
        ANHTTPStubbingManager.shared().broadcastRequests = false
        ANSDKSettings.sharedInstance().httpsEnabled = false
        NotificationCenter.default.removeObserver(self)
        self.loadAdSuccesfulException = nil
    }
    
    func requestCompleted(_ notification: Notification?) {
        var incomingRequest = notification?.userInfo![kANHTTPStubURLProtocolRequest] as? URLRequest
        let requestString = incomingRequest?.url?.absoluteString
        let searchString = ANSDKSettings.sharedInstance().baseUrlConfig.utAdRequestBaseUrl()
        if request == nil && requestString?.range(of:searchString!) != nil{
            request = notification!.userInfo![kANHTTPStubURLProtocolRequest] as! URLRequest
            jsonRequestBody = ANHTTPStubbingManager.jsonBodyOfURLRequest(asDictionary: request) as! [String : Any]
        }
    }
    
    //MARK:- Test Methods
    //Test ANGDPRSettings for GDPR consent to true
    func test_TC58_ANGDPRSettingsForSetGDPRConsentTrue() {
        ANGDPRSettings.setConsentRequired(true)
        ANGDPRSettings.setConsentString("a390129402948384453")
        self.banner = ANBannerAdView.init(frame: CGRect(x: 0, y: 0, width: 300, height: 250), placementId: placementID, adSize: CGSize(width: 300, height: 250))
        self.banner.delegate = self
        stubRequestWithResponse("SuccessfulInstreamVideoAdResponse")
        loadAdSuccesfulException = expectation(description: "\(#function)")
        banner.loadAd()
        waitForExpectations(timeout: timeoutForImpbusRequest, handler: nil)
        if let dic = jsonRequestBody["gdpr_consent"] as? [String : Any], let consent_required = dic["consent_required"] as? Bool
        {
            XCTAssertTrue(consent_required)
        }
        if let dic = jsonRequestBody["gdpr_consent"] as? [String : Any], let consent_string = dic["consent_string"] as? String
        {
            XCTAssertNotNil(consent_string)
        }
    }
    
    //Test ANGDPRSettings for GDPR consent to false
    func test_TC59_ANGDPRSettingsForSetGDPRConsentFalse() {
        ANGDPRSettings.setConsentRequired(false)
        ANGDPRSettings.setConsentString("a390129402948384453")
        self.banner = ANBannerAdView.init(frame: CGRect(x: 0, y: 0, width: 300, height: 250), placementId: placementID, adSize: CGSize(width: 300, height: 250))
        self.banner.delegate = self
        stubRequestWithResponse("SuccessfulInstreamVideoAdResponse")
        loadAdSuccesfulException = expectation(description: "\(#function)")
        banner.loadAd()
        waitForExpectations(timeout: timeoutForImpbusRequest, handler: nil)
        if let dic = jsonRequestBody["gdpr_consent"] as? [String : Any], let consent_required = dic["consent_required"] as? Bool
        {
            XCTAssertFalse(consent_required)
        }
        if let dic = jsonRequestBody["gdpr_consent"] as? [String : Any], let consent_string = dic["consent_string"] as? String
        {
            XCTAssertNotNil(consent_string)
        }
    }
    
    //Test ANGDPRSettings for default GDPR Consent
    func test_TC60_ANGDPRSettingsForSetGDPRDefaultConsent() {
        ANGDPRSettings.reset()
        self.banner = ANBannerAdView.init(frame: CGRect(x: 0, y: 0, width: 300, height: 250), placementId: placementID, adSize: CGSize(width: 300, height: 250))
        self.banner.delegate = self
        stubRequestWithResponse("SuccessfulInstreamVideoAdResponse")
        loadAdSuccesfulException = expectation(description: "\(#function)")
        banner.loadAd()
        waitForExpectations(timeout: timeoutForImpbusRequest, handler: nil)
        XCTAssertNil(jsonRequestBody["gdpr_consent"] )
    }
    
    //Test ANGDPRSettings for request content type
    func test_TC61_ANGDPRSettingsRequestContentType() {
        self.banner = ANBannerAdView.init(frame: CGRect(x: 0, y: 0, width: 300, height: 250), placementId: placementID, adSize: CGSize(width: 300, height: 250))
        self.banner.delegate = self
        stubRequestWithResponse("SuccessfulInstreamVideoAdResponse")
        loadAdSuccesfulException = expectation(description: "\(#function)")
        banner.loadAd()
        waitForExpectations(timeout: timeoutForImpbusRequest, handler: nil)
        let contentType = request.value(forHTTPHeaderField: "content-type")
        XCTAssertNotNil(contentType)
        XCTAssertEqual("application/json", contentType)
    }
    
    // MARK: - Stubbing
    func stubRequestWithResponse(_ responseName: String?) {
        let currentBundle = Bundle(for: type(of: self))
        let baseResponse = try? String(contentsOfFile: currentBundle.path(forResource: responseName, ofType: "json") ?? "", encoding: .utf8)
        let requestStub = ANURLConnectionStub()
        requestStub.requestURL = ANSDKSettings.sharedInstance().baseUrlConfig.utAdRequestBaseUrl()
        requestStub.responseCode = 200
        requestStub.responseBody = baseResponse
        
        ANHTTPStubbingManager.shared().add(requestStub)
    }
    
    @objc func fulfillExpectation(_ expectation: XCTestExpectation?) {
        expectation?.fulfill()
    }
    
    func waitForTimeInterval(_ delay: TimeInterval) {
        let expectation: XCTestExpectation = self.expectation(description: "wait")
        perform(#selector(self.fulfillExpectation(_:)), with: expectation, afterDelay: delay)
        
        waitForExpectations(timeout: TimeInterval(delay + 1), handler: nil)
    }
    
    // MARK: - ANAdDelegate
    func adDidReceiveAd(_ ad: Any!) {
        loadAdSuccesfulException?.fulfill()
    }
    
    func ad(_ ad: Any!, requestFailedWithError error: Error!) {
        loadAdSuccesfulException?.fulfill()
    }
}
